streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1073,32 +1071,26 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
    */
   final def mapMPar[R1 <: R, E1 >: E, B](n: Int)(f: A => ZIO[R1, E1, B]): ZStream[R1, E1, B] =
     new ZStream[R1, E1, B] {
-      def fold[R2 <: R1, E2 >: E1, B1 >: B, S]: ZStream.Fold[R2, E2, B1, S] =
-        ZManaged.succeed { (s, cont, g) =>
-          for {
-            out              <- Queue.bounded[Take[E1, IO[E1, B]]](n).toManaged(_.shutdown)
-            permits          <- Semaphore.make(n.toLong).toManaged_
-            interruptWorkers <- Promise.make[Nothing, Unit].toManaged_
-            _ <- self.foreachManaged { a =>
-                  for {
-                    p <- Promise.make[E1, B]
-                    _ <- out.offer(Take.Value(p.await))
-                    _ <- (permits.withPermit(f(a).to(p)) race interruptWorkers.await).fork
-                  } yield ()
-                }.foldCauseM(
-                    c => (out.offer(Take.Fail(c)) *> ZIO.halt(c)).toManaged_,
-                    _ => out.offer(Take.End).unit.toManaged_
-                  )
-                  .ensuring(interruptWorkers.succeed(()) *> permits.withPermits(n.toLong)(ZIO.unit))
-                  .fork
-            s <- Stream
-                  .fromQueue(out)
-                  .unTake
-                  .mapM(identity)
-                  .fold[R2, E2, B1, S]
-                  .flatMap(fold => fold(s, cont, g))
-          } yield s
-        }
+      def fold[R2 <: R1, E2 >: E1, B1 >: B, S]: ZStream.Fold[R2, E2, B1, S] = foldDefault
+
+      override def process =
+        for {
+          out              <- Queue.bounded[InputStream[R1, E1, B]](n).toManaged(_.shutdown)
+          permits          <- Semaphore.make(n.toLong).toManaged_
+          interruptWorkers <- Promise.make[Nothing, Unit].toManaged_
+          _ <- self.foreachManaged { a =>
+                for {
+                  p <- Promise.make[E1, B]
+                  _ <- out.offer(InputStream.fromPromise(p))
+                  _ <- (f(a).to(p) race interruptWorkers.await).fork
```

Comment:
`withPermit` is missing here on the `f(a).to(p)`

Reply:
Yes, mistake when reverting code...